### PR TITLE
Aws iam

### PR DESCRIPTION
**Pull request by eyalpalo**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://github.com/demisto/etc/issues/40873

## Description
AWS IAM integration enhancements. Four commands added.

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 1]
```diff
@@ -2277,9 +2291,182 @@ script:
...
+      type: String
+    - contextPath: AWS.IAM.AttachedGroupPolicies.PolicyName
+      description: The name of the attached policy
+      type: String
+    - contextPath: AWS.IAM.AttachedGroupPolicies.PolicyArn
+      description: The Amazon Resource Name (ARN) of the attached policy.
+      type: String
+    - contextPath: AWS.IAM.Groups.AttachedPoliciesMarker
+      description: First element of next page of items.
+      type: Unknown
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2449]:
```suggestion
      type: String
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    def list_attached_group_policies(self):
+        pass
+
+    def get_login_profile(self):
+        pass
+
+
+@pytest.mark.parametrize('args, res', PAGINATION_CHECK)
+def test_list_user_policies(mocker, args, res):
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py:56]: please add docstrings to all test functions ("given, when, then")

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
+import pytest
+import importlib
+import demistomock as demisto
+from CommonServerPython import tableToMarkdown, pascalToSpace
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py:4]:
```suggestion
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    mocker.patch.object(demisto, 'results')
+
+    client = AWSClient()
+    AWS_IAM.list_user_policies(args, client)
+    contents = demisto.results.call_args[0][0]
+    human_readable = tableToMarkdown('AWS IAM Policies for user {}'.format('test'),
+                                     headers=["PolicyNames"],
+                                     headerTransform=pascalToSpace,
+                                     t=res)
+    assert contents.get('HumanReadable') == human_readable
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py:88]:
```suggestion
    assert 'AWS IAM Policies for user test' in contents.get('HumanReadable')
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    ec = {'AWS.IAM.AttachedUserPolicies(val.PolicyArn && val.UserName && val.PolicyArn === obj.PolicyArn && '
+          'val.UserName === obj.UserName)': [{'UserName': 'test', 'PolicyName': policy_name, 'PolicyArn': policy_arn}],
+          'AWS.IAM.Users(val.UserName === \'{}\').AttachedPoliciesMarker'.format('test'): '111'}
+
+    human_readable = tableToMarkdown('AWS IAM Attached Policies for user {}'.format('test'),
+                                     headers=['PolicyName', 'PolicyArn'],
+                                     headerTransform=pascalToSpace,
+                                     t=ATTACHED_POLICIES)
+    assert contents.get('HumanReadable') == human_readable
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py:116]: same here

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                                     headerTransform=pascalToSpace,
+                                     t=ATTACHED_POLICIES)
+
+    ec = {'AWS.IAM.AttachedGroupPolicies(val.PolicyArn && val.GroupName && val.PolicyArn === obj.PolicyArn && '
+          'val.GroupName === obj.GroupName)':
+              [{'GroupName': 'test', 'PolicyName': policy_name, 'PolicyArn': policy_arn}],
+          'AWS.IAM.Groups(val.GroupName === \'{}\').AttachedPoliciesMarker'.format('test'): '111'}
+
+    assert contents.get('HumanReadable') == human_readable
+    assert contents.get('EntryContext') == ec
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py:147]: same here

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    policy_data = []
+    for policy in res:
+        policy_data.append({
+            'UserName': 'test',
+            'PolicyName': policy,
+        })
+
+    ec = {'AWS.IAM.UserPolicies(val.PolicyName && val.UserName && val.PolicyName === obj.PolicyName && '
+          'val.UserName === obj.UserName)': policy_data,
+          'AWS.IAM.Users(val.UserName === \'{}\').InlinePoliciesMarker'.format('test'): response.get('Marker')}
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS_IAM_test.py:75]: you are pretty much providing the same code lines as the function you are testing, no need to do all of this. Just check if the mock data exists in the response.

**Review by dantavori — CHANGES_REQUESTED** [review]

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py [hunk 8]
```diff
@@ -20,6 +20,22 @@ def default(self, obj):
         return json.JSONEncoder.default(self, obj)
 
 
+def get_limit(args):
+    limit = arg_to_number(str(args.get("limit"))) if "limit" in args else None
+    page = arg_to_number(str(args.get("page"))) if "page" in args else None
+    page_size = arg_to_number(str(args.get("page_size"))) if "page_size" in args else None
```
[the remark this answers is not in the document]
Comment (eyalpalo) [reply Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py:26]: The reason I used casting to str is the fact that the args returning from the get function are of type unicode and not str or int. arg_to_number fails if the args is not of type str or int.

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py [hunk 9]
```diff
@@ -1003,6 +1019,174 @@ def get_policy(args, aws_client):
...
+        role_session_name=args.get('roleSessionName'),
+        role_session_duration=args.get('roleSessionDuration'),
+    )
+    user_name = args.get('userName', "")
+    marker = args.get('marker', None)
+    limit, is_manual, page_size = get_limit(args)
+
+    kwargs = {
+        'UserName': user_name,
+        'MaxItems': limit
```
[the remark this answers is not in the document]
Comment (eyalpalo) [reply Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py:1035]: Marker only marks the first object from which to start the response. if the user uses limit and marker I used the MaxItems to return only limit items starting the marker.

**Review by dantavori — COMMENTED** [review]
looks good to me, will approve after the demo

**Review by dantavori — APPROVED** [review]

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 10]
```diff
@@ -2277,9 +2291,182 @@ script:
...
+  - arguments:
+    - default: false
+      description: The name (friendly name, not ARN) of the user to list attached
+        policies for.
+      isArray: false
+      name: userName
+      required: true
+      secret: false
+    - default: false
+      description: Number of results to display. Default value is 50.
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2303]:
```suggestion
      description: Number of results to display.
      defaultValue: '50'
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 11]
```diff
@@ -2277,9 +2291,182 @@ script:
...
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2353]:
```suggestion
      description: Number of results to display.
      defaultValue: '50'
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 12]
```diff
@@ -2277,9 +2291,182 @@ script:
...
+  - arguments:
+    - default: false
+      description: The name (friendly name, not ARN) of the group to list attached
+        policies for.
+      isArray: false
+      name: groupName
+      required: true
+      secret: false
+    - default: false
+      description: Number of results to display. Default value is 50.
```
Comment (dantavori) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2406]:
```suggestion
      description: Number of results to display.
      defaultValue: '50'
```

**Review by dantavori — APPROVED** [review]

**Pull request merged**